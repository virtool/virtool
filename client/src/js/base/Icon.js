import React, { useCallback } from "react";
import PropTypes from "prop-types";
import { capitalize, get } from "lodash-es";
import { Link } from "react-router-dom";
import styled, { css } from "styled-components";
import { Tooltip } from "./Tooltip";

const getIconColor = ({ color, theme, shade = "dark" }) =>
    get(theme, ["color", `${color}${capitalize(shade)}`], "inherit");

const fixedWidth = css`
    width: 8px;
    text-align: center;
    display: inline-block;
`;

const StyledIcon = styled.i`
    color: ${getIconColor};
    ${props => (props.hoverable || props.onClick ? "cursor: pointer;" : "")};
    opacity: ${props => (props.hoverable || props.onClick ? 0.7 : 1)};

    ${props => (props.fixedWidth ? fixedWidth : "")};

    :hover {
        opacity: 1;
    }
`;

export const Icon = ({ hoverable, style, ...props }) => {
    const handleClick = useCallback(e => {
        props.onClick(e);
    }, []);

    const className = `${props.className ? props.className + " " : ""} ${props.faStyle} fa-${props.name}`;

    const icon = (
        <StyledIcon
            className={className}
            fixedWidth={props.fixedWidth}
            hoverable={hoverable}
            style={style}
            onClick={props.onClick ? handleClick : null}
            color={props.color}
            shade={props.shade}
        />
    );

    if (props.tip) {
        return (
            <Tooltip position={props.tipPlacement || "top"} tip={props.tip}>
                {icon}
            </Tooltip>
        );
    }

    return icon;
};

Icon.propTypes = {
    color: PropTypes.oneOf(["blue", "green", "grey", "red", "orange", "purple"]),
    name: PropTypes.string.isRequired,
    tip: PropTypes.node,
    tipPlacement: PropTypes.oneOf(["top", "right", "bottom", "left"]),
    faStyle: PropTypes.string,
    onClick: PropTypes.func,
    className: PropTypes.string,
    fixedWidth: PropTypes.bool,
    style: PropTypes.object
};

Icon.defaultProps = {
    faStyle: "fas",
    fixedWidth: false
};

export const LinkIcon = ({ to, replace, ...props }) => (
    <Link to={to} replace={replace} hoverable>
        <Icon {...props} />
    </Link>
);
