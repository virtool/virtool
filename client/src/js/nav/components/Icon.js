import React from "react";
import { connect } from "react-redux";
import { Overlay } from "react-bootstrap";
import Notifications from "./Notifications";
import { getSoftwareUpdates } from "../../updates/actions";
import { getUnbuilt } from "../../indexes/actions";

import { Icon } from "../../base";

const hrToMs = 3600000;

class NotificationIcon extends React.Component {

    constructor (props) {
        super(props);

        this.state = {
            show: false
        };
    }

    handleToggle = () => {

        if (this.state.show) {
            window.removeEventListener("click", this.handleExit, false);
        } else {
            window.addEventListener("click", this.handleExit, false);
        }

        this.setState({
            show: !this.state.show
        });
    }

    handleExit = (e) => {
        if (!this.target.contains(e.target)) {
            this.handleToggle();
        }
    }

    componentWillMount () {
        this.props.onGet();
        this.props.onGetUnbuilt();
    }

    componentDidMount () {
        this.interval = window.setInterval(() => {
            this.props.onGet();
            this.props.onGetUnbuilt();
        }, (3 * hrToMs));
    }

    componentWillUnmount () {
        window.clearInterval(this.interval);
    }

    render () {

        const availableUpdates = this.props.updates ? this.props.updates.releases.length : null;
        const unbuiltIndex = ((this.props.modifiedVirusCount || this.props.modifiedCount) && this.props.unbuilt)
            ? this.props.unbuilt.history.length
            : null;

        const iconStyle = (availableUpdates || unbuiltIndex) ? "icon-pulse" : "icon";

        return (
            <div>
                <div ref={node => this.target = node} onClick={this.handleToggle}>
                    <Icon
                        className={iconStyle}
                        name="notification"
                        tip="Click to see notifications"
                        tipPlacement="left"
                    />
                </div>

                <Overlay
                    show={this.state.show}
                    placement="bottom"
                    target={this.target}
                >
                    <Notifications onClick={this.handleToggle} />
                </Overlay>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    updates: state.updates.software,
    unbuilt: state.indexes.unbuilt,
    modifiedVirusCount: state.indexes.modified_virus_count,
    modifiedCount: state.viruses.modified_count
});

const mapDispatchToProps = (dispatch) => ({

    onGet: () => {
        dispatch(getSoftwareUpdates());
    },

    onGetUnbuilt: () => {
        dispatch(getUnbuilt());
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(NotificationIcon);
