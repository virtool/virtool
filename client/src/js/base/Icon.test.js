import { Icon } from "./Icon";
import { Tooltip, OverlayTrigger } from "react-bootstrap";

describe("<Icon />", () => {
    let props;
    let wrapper;

    it("renders correctly", () => {
        props = {
            name: "i-test"
        };
        wrapper = shallow(<Icon {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("adds 3px margin-left property when [props.pad=true]", () => {
        props = {
            name: "i-test",
            pad: true
        };

        wrapper = shallow(<Icon {...props} />);
        expect(wrapper.find('i').prop('style')).toEqual({ marginLeft: "3px" });
    });

    it("adds spinner css class when [props.pending=true]", () => {
        props = {
            name: "i-test",
            pending: true
        };
        wrapper = shallow(<Icon {...props} />);

        expect(wrapper.find('i').hasClass("i-spinner spinning")).toBe(true);
    });

    it("add bsStyle class when bsStyle prop is given and [prop.pending=false]", () => {
        props = {
            name: "i-test",
            pending: false,
            bsStyle: "danger"
        };
        wrapper = shallow(<Icon {...props} />);

        expect(wrapper.find('i').hasClass(`text-${props.bsStyle}`)).toBe(true);
    });

    describe("when props.tip is supplied", () => {

        it("renders Tooltip + OverlayTrigger components", () => {
            props = {
                name: "i-test",
                tip: "test_tip"
            };
            wrapper = mount(<div><Icon {...props} /></div>);

            expect(wrapper.find(OverlayTrigger).length).toEqual(1);

            const target = wrapper.find(OverlayTrigger).prop('overlay');

            expect(target.type.name).toEqual("Tooltip");
            expect(target.props.children).toEqual(props.tip);
        });

        it("renders accordingly to props.tipPlacement (if supplied)", () => {
            props = {
                name: "i-test",
                tip: "test tip",
                tipPlacement: "right"
            };
    
            wrapper = shallow(<Icon {...props} />);
            expect(wrapper.find(OverlayTrigger).prop('placement')).toEqual(props.tipPlacement);
        });

    });

    describe("when props.tip is not supplied", () => {

        it("renders <i> element", () => {
            props = {
                name: "i-test"
            };
            wrapper = shallow(<Icon {...props} />);

            expect(wrapper.length).toEqual(1);
            expect(wrapper.find('i').exists()).toBe(true);
        });

        it("is clickable if onClick is supplied", () => {
            props = {
                name: "i-test",
                onClick: jest.fn()
            };
            wrapper = shallow(<Icon {...props} />);

            wrapper.find('i').simulate('click', { stopPropagation: () => {} });

            expect(props.onClick).toHaveBeenCalled();
        });
    });

});
