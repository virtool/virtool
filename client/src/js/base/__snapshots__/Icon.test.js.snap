// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> renders correctly 1`] = `
<i
  className="i-i-test"
  onClick={null}
  style={Object {}}
/>
`;
